### PR TITLE
Time - Vicki 

### PR DESCRIPTION
**Pull request by vwhwang**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | top_work and top_10 to show highest votes media
Describe how you approached testing that model method. What edge cases did you come up with? | I didn't try the edge cases 
What are `session` and `flash`? What is the difference between them? | sessions will log info for the browser and it will last via the session. while flash last for the request action.
What was one thing that you gained more clarity on through this assignment? | relationship of tables 
What is the Heroku URL of your deployed application? | https://vicki-media-ranker.herokuapp.com/

### CODE REVIEW

File: app/controllers/works_controller.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @work = Work.find_by(id: params[:id])
+
+    if @work.nil?
+      head :not_found
+      return 
+    end 
+  end 
+
+  def update
+    @work = Work.find_by(id: params[:id])
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:54]: Consider putting this functionality into a controller filter.

File: test/models/work_test.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      new_work = Work.create(title: work.title)
+      result = new_work.valid? 
+
+      expect(result).must_equal false 
+      expect(new_work.errors.messages[:title].include?("has already been taken")).must_equal true 
+
+    end 
+
+    describe "top logics " do 
+      it "top_work will create one top work" do
```
Comment (beccaelenzil) [inline test/models/work_test.rb:38]: Consider edge cases. For instance --What about tied works? What if there are no works in a particular category?

**Comment by beccaelenzil** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | These are not sorted by vote.
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️redirect
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️No, but you do have `require_login` :)
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | See in-line comment.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | As you mentioned, you did not test edge cases. See in-line comment for suggestions.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project. It is clear that the learning objectives were met. Your code is clear and well written. I've left a few comments about how you could make use of controller filters and more thoroughly test your model. Let me know if you have any questions. Keep up the hard work!
